contract/src/lib.rs

```diff
@@ -253,22 +287,31 @@ impl BtcLightClient {
             ));
 
             for height in start_removal_height..end_removal_height {
-                let blockhash = &self.mainchain_height_to_header[&height];
+                let blockhash = &self
+                    .mainchain_height_to_header
+                    .get(&height)
+                    .unwrap_or_else(|| env::panic_str(ERR_KEY_NOT_EXIST));
 
                 self.headers_pool.remove(blockhash);
                 self.mainchain_header_to_height.remove(blockhash);
                 self.mainchain_height_to_header.remove(&height);
             }
 
-            self.mainchain_initial_blockhash
-                .clone_from(&self.mainchain_height_to_header[&end_removal_height]);
+            self.mainchain_initial_blockhash = self
+                .mainchain_height_to_header
+                .get(&end_removal_height)
+                .unwrap_or_else(|| env::panic_str(ERR_KEY_NOT_EXIST));
         }
     }
 }
 
 impl BtcLightClient {
-    fn init_genesis(&mut self, block_header: Header, block_height: u64) {
+    fn init_genesis(&mut self, block_header: Header, block_hash: &H256, block_height: u64) {
```

Comment:
Why do we need to pass the hash if it can be easily calculated from the header?

Reply:
This prevents initializing contact with the wrong header data; this is happened to me twice on testnet.